### ISSUE TITLE
terraform init seems to not support basic terraform best practice

### ISSUE THREAD

**Issue by gtmtech** [hashicorp/terraform#18632]
Terraform all versions. 

It's terraform best practice to store environment config in tfvars files . e.g:

```
    environments/
        |-- dev.tfvars      # store dev configuration
        |-- prod.tfvars    # store stage configuration
    resources/
        |-- main.tf           # resources etc.
```

(Nice clean separation of environment config from other config from terraform resources - good!)

And to use e.g. `terraform plan -var-file environments/dev.tfvars`

But (in terms of AWS at least) its also best practice to separate dev, prod in terms of aws accounts (for one, you like to have business/enterprise support on prod which is charged as a %age of complete account cost, so having everything in the same account is very expensive for support; for another its good for security boundaries anyway as some security fundamentals can only be applied on an account basis). This yields a natural fit of storing state in a different S3 bucket for each environment too - eg `terraform-dev` in dev, `terraform-prod` in prod. (If you dont do this, then you have far more cross-account assume-roles to do, because the state has to live in one account, whilst the terraforming is done in another account, but even if you do store in the same bucket the following still applies...)

Up to now, all fine and nice and clean, but terraform init doesn't cope with this at all. Why? because terraform init is rigid and has to use .terraform locally to store its local version of a single state.

Therefore a backend that looks like this:

```
terraform {
  backend "s3" {
#   The bucket cannot be specified here as it changes between environments
    dynamodb_table = "terraform_lock"
    encrypt        = true
    key            = "main.tfstate"
    region         = "eu-west-1"
  }
}
```

Should work in conjunction with `terraform init -no-input -backend-config=bucket=terraform-dev` Likewise it should work when "switching" with `terraform init -no-input -backend-config=bucket=terraform-prod`

But it's made needlessly complicated because firstly the second terraform init needs to see both buckets if you dont add `-reconfigure` (For some reason, I guess it wants to support state migration?):

```
Prior to changing backends, Terraform inspects the source and destination
states to determine what kind of migration steps need to be taken, if any.
Terraform failed to load the states. The data in both the source and the
destination remain unmodified. Please resolve the above error and try again.
```

and secondly, with `-reconfigure`, it introduces a huge delay anyway as downloading modules takes a while, and syncing/reading the state each time from a different place takes a while, and so changing the environments makes for slow and frustrating terraform experience.

Stored state should be in different files or buckets for each environment, otherwise anyone read/writing dev state can also read/write prod state and you have no graduated security in your team (which is fine in some places but not all). But constantly having to redo the terraform init every time you switch from dev to prod is a complete pain. Workspaces are also not the answer to this problem (once again storing stuff in the same place?)

Something as simple as allowing different places for the .terraform directory would completely fix this. ( #3503 ) I could then init twice - one for dev, one for prod, store them both and be a happy terraformer just by means of an environment variable- Not only this, but it brings so many advantages because it allows you to run multiple terraforms concurrently against different environments on the same source tree (terraform init completely blocks this having a single place). 

This seems like such a trivial enhancement, can this be fixed? Or can others elaborate on what they do? Maybe I'm missing something obvious. My requirements are:

* Store state remotely in different places for each environment for security reasons
* Not have to run terraform init every time I switch environments as it takes ages

with a bonus (not requirement) that:

* I can run terraform in dev and prod concurrently off the same working tree

Right now I work around this with symlinks and copying the whole working directory to a tmp location. It really sucks. But environments are a fundamental part of terraform so what am I missing?

Thanks!

@apparentlymart you might be interested in this?

**Comment by myungmobile** [comment]
+1

**Comment by gtmtech** [comment]
@apparentlymart Thanks for taking the time for a detailed explanation! I have seen the directory per environment approach used above, and appreciate the comments you make about it.

In my experience this approach leads to teams having 2 places to put resources - in the (common) resources folder, or in the (environment specific) folder. Over time teams can put more and more in the environment specific folders (as things may initially just be for one environment) and they end up with a difficult-to-compare set of resources and possibly also multiple copy+pastes across different environments with the inevitable question (why isnt production anything like staging) question being asked later down the line.

It's for this reason I prefer instead to mandate a single resources area, and use feature-flags (a variable map "flags") and copious use of `count` on all resources to toggle features on and off in environments, and use tfvars lists and maps to wire in other config. In this way the differences between two environments are just comparing two tfvars files which I find simpler for team to get a handle on, than having to compare trees of differently configured resources and submodules, and with a single area for resources, teams have to make sure they work across all environments, which means they dont end up putting any environment specific resources in the resource definitions themselves. This tends to lead to cleaner, DRYer code.

However I also appreciate that your approach solves the immediate problem I'm having of separate state files (although it also requires me to cd quite a lot). 

I think on balance if you're up for a PR, I might see if I can do one, to use an optional env var for the location of the .terraform directory, and then we can use both approaches in different situations where they make sense for the end users?

**Comment by apparentlymart** [comment]
Thanks for that extra context, @gtmtech!

It is true that the module-per-environment approach does rely on humans to police themselves and each other (complying with a policy on which resources -- if any -- belong in the environment-specific module) whereas using variables implicitly forces that through the limitations of a `.tfvars` file. Our usual attitude is that Terraform should encourage a best-practice but give you room to step out of it when it's not appropriate, and so we generally lean towards the "trust the humans" angle when designing features.

However, given that moving the `.terraform` directory doesn't seem like it would create any particular _harm_ (either directly for end-users or making future development more difficult) I'm happy to be flexible and make room for this alternative approach, even though I don't feel ready to call it a "best practice" yet. (Good experience reports may change my mind! :grinning:)

I think it is probably best to wait until the long-running 0.12 development branch is merged into master before starting a PR on this, since I expect they'll cover similar ground and we'd end up having to resolve merge conflicts down the road otherwise. However, once that is true (still gonna be at least a few weeks away though, I expect :confounded:) we'd be happy to review a PR for this.

Thanks again for writing this up, and for the interest in working on a PR!

**Comment by davidgoate** [comment]
@apparentlymart A couple of questions about the recommended structure above:


```
    environments/
        |-- dev/ # dev configuration
            |-- dev.tf
        |-- prod/ # prod configuration
            |-- prod.tf
    resources/ # shared module for elements common to all environments
        |-- main.tf
```

using example `dev.tf`

```yaml
module "main" {
  source = "../resources"

  # per-environment settings here, as you had in the .tfvars files in your example
}
```

1. Is the first source attribute meant to be `../../resources` instead?
1. How do outputs work in this structure? I used to have an `outputs.tf` in the root directory and now I see no output when running 

>terraform output
The state file either has no outputs defined, or all the defined
outputs are empty. Please define an output in your configuration
with the `output` keyword and run `terraform refresh` for it to
become available. If you are using interpolation, please verify
the interpolated value is not empty. You can use the 
`terraform console` command to assist.

My old structure was:

```
        |-- file1.tf
        |-- file2.tf
        |-- variables.tf
        |-- outputs.tf
```

Now I have:

```
    environments/
        |-- dev/ 
            |-- dev.tf
        |-- prod/ 
            |-- prod.tf
    resources/ # Everything that used to be at the root directory level
         |-- file1.tf
         |-- file2.tf
         |-- variables.tf
         |-- outputs.tf
```

My dev.tf looks like this:


```yaml
terraform {
  backend "s3" {
    ...
  }
  version = "~> 0.11.8"
}

provider "aws" {
...
}

module "main" {
  source = "../../resources"
}
```

When I run from the dev directory `terraform apply` or `terraform output` I see no outputs, is there a way to "import" the `outputs.tf` so that all output are printed?

**Comment by davidgoate** [comment]
@apparentlymart I think I figured out a way to get the outputs in the structure, from within the `dev` directory I now run `terraform output --module=main`

**Comment by abeluck** [comment]
Does there exist a repo on github (or elsewhere) that demonstrates this layout by @apparentlymart (either an example repo or a real-thing)?

I'm diving into a new AWS project as a newbie, and while trawling for info on how to organize my terraform code while handling multiple environments I ran across this issue.

**Comment by apparentlymart** [comment]
I think a general challenge with showing "full examples" for something like this is that they necessarily involve making some decisions that are not really relevant to the pattern and can distract from the point being made. Defining environments in Terraform involves deciding what an environment is "made of" in your organization, which is a subjective architectural decision separate from how the Terraform configurations for representing those environments are organized.

With that said, in a prior role (before joining the Terraform team at HashiCorp) I wrote up an overview of a generalization of that organization's structure in my personal blog: [Creating Environments](https://apparently.me.uk/terraform-environment-application-pattern/creating-environments.html). It's part of [a bigger series of posts](https://apparently.me.uk/terraform-environment-application-pattern/) that describe the overall approach, but that particular article is about codifying the environments themselves in Terraform. That content is from 2017, so some details may have shifted in the mean time, but I think the general approach is still valid.

**Comment by oliverleach** [comment]
I've been sent this issue and after reviewing, I'm don't agree with the below statement.

> I'd actually recommend a different layout as a "**best practice**":
> 
> ```
>     environments/
>         |-- dev/ # dev configuration
>             |-- dev.tf
>         |-- prod/ # prod configuration
>             |-- prod.tf
>     resources/ # shared module for elements common to all environments
>         |-- main.tf
> ```

What this is suggesting is managing multiple tf files per environment. This being stated as best practice is a concern as it really seems to me to be more of a work around to a missing piece of functionality that terraform doesn't provide, as stated by the original author @gtmtech.

My main issue with this approach is that I now have to keep in sync multiple <env>.tf files and this doesn't follow with the principles of abstraction. Keeping different files for different environments in sync is a big concern, especially as my environments grow and the files get more complicated to keep in sync. As my environments get more complex, my references to modules also increase as new functionality come in. We should abstract the environment specifics and parameterize the terraform file, passing in environment variables in the form of tfvar file. 

@apparentlymart, are we sure this is best practice? 

Ultimately, using this best practice advice, there is no guarantee that what gets tested and deployed in dev uses the same deployment files in prod as, actually, they are different files. 

Thoughts anyone?

**Comment by apparentlymart** [comment]
You can mitigate that concern by making sure that each environment module contains only a `backend` configuration and a single `module` block with the same source in each environment. Then the environment modules contain only the information that is actually different between the environments, and the shared module contains the common elements.

How similar the environments are is ultimately up to you.

**Comment by gtmtech** [comment]
@apparentlymart  - this is indeed what I do now, however its a bit more complicated than that - 

You need to have at a minimum, `backend.tf`, single `module` block, `variable` declarations and `provider` declarations. Hmm, well I guess you can define `providers within the module actually, but the variables are needed. 

This can be condensed into a single file, but the previous tfvars collections of file look and feel so much better, I've gone back to using them this way, and wrapping terraform again. 

I think when I get time Ill work on that support for storing tempfiles in a user-specified directory rather than `.terraform` as this will make it a lot simpler to do stuff at scale.
